### PR TITLE
Add version compatibility checks for extensions

### PR DESCRIPTION
**Pull request by blink1073**
Use npm semver to check for lab extension validity.

Adds version compatibility to lab extension validation and does not attempt to load invalid entries.

Also fixes our Python version to be in sync with npm (it was just `None`) before, and allows `jupyter lab --version` to print the appropriate version.

This will only work for plugins created using `@jupyterlab/extension-builder` 0.8.1 or newer, since we were previously not adding the required dependencies metadata for each module in the manifest.

```
jupyter labextension list
Known labextensions:
config dir: /Users/ssilvester/.jupyter/labconfig
    jupyterlab_geojson  enabled
    - Validating...
     OK has entry point?
     OK has necessary files?
      X is version compatible?
```

cf #1011.

### CODE REVIEW

**Comment by Carreau** [comment]
I'm just going to note that python version are supposed to adhere to [pep440](https://www.python.org/dev/peps/pep-0440/) and that pip/PyPI do make comparison to decide wether packages need to be upgraded base on that, on not on semver. This is (in practice) the same, except in beta/rc cycle and with dev versions, where suffix does matter, and in particular the dot between the patch number and dev/b/a suffix can lead to surprising behavior. For example a missing (or extra) dot can lead to the python wheel and .tar.gz being treated as 2 different version of the package, which can lead to some confusions.

I've made [a pep440](https://pypi.python.org/pypi/pep440) that expose a "iscanonical" function to make sure a release number won't be problematic with Python/pip.

**Comment by jasongrout** [comment]
> I'm just going to note that python version are supposed to adhere to pep440 and that pip/PyPI do make comparison to decide wether packages need to be upgraded base on that, on not on semver. This is (in practice) the same, except in beta/rc cycle and with dev versions, where suffix does matter, and in particular the dot between the patch number and dev/b/a suffix can lead to surprising behavior. For example a missing (or extra) dot can lead to the python wheel and .tar.gz being treated as 2 different version of the package, which can lead to some confusions.

Thanks. We don't use prerelease versions because semver has lots of problems with them too.

**Comment by jasongrout** [comment]
The algorithm here won't quite work. For example, using the extension builder 0.8.1 to build jupyterlab_geojson generates this in the manifest:

```
        "modules": {
                "jupyterlab_geojson@0.1.1/lib/plugin.js": [
                        "jupyterlab@^0.7.0/lib/docregistry/index.js",
                        "jupyterlab_geojson@0.1.1/lib/widget.js",
                        "leaflet@^1.0.1/dist/leaflet.css"
                ],
                "jupyterlab@0.7.0/lib/docregistry/index.js": [
                        "jupyterlab@0.7.0/lib/docregistry/context.js",
                        "jupyterlab@0.7.0/lib/docregistry/default.js",
                        "jupyterlab@0.7.0/lib/docregistry/kernelactions.js",
                        "jupyterlab@0.7.0/lib/docregistry/kernelselector.js",
                        "jupyterlab@0.7.0/lib/docregistry/registry.js"
                ],
```

Now, this extension should be good with jlab `0.7.1`, even if it includes `0.7.0`. But when the algorithm hits the list of dependencies for the jlab module, it will try to see if `0.7.1` satisfies the semver range `0.7.0`.

Some options:
1. Ignore any mods that were `jupyterlab`, knowing that the semver range of those dependencies is restricted to that specific version.
2. Add a separate `engine` attribute to the manifest saying what semver range of jlab is appropriate, pulled straight from the package.json. This would require a change to the extension builder, of course.
3. Look at the manifest's `entry` attribute to see what mods to check, and ignore the others. (I think this option is not good by itself, because it doesn't check dependency's versions of jlab, but it might be helpful to flag to the user if a dependency actually depends on a different version of jlab)

But even all of these, while improvements on the status quo, don't address the real problem of modules provided in a package shadowing the system version. For example, suppose my plugin packages jlab 0.7.1 modules, but I'm trying to run it on 0.7.0. The semver cache will still serve 0.7.1 modules to everyone from my extension.

**Comment by jasongrout** [comment]
> 1. Ignore any mods that were jupyterlab, knowing that the semver range of those dependencies is restricted to that specific version.

Minimally, I think this can go in, and is an improvement, if the above is done.

**Comment by jasongrout** [comment]
> > Ignore any mods that were jupyterlab, knowing that the semver range of those dependencies is restricted to that specific version.
> Minimally, I think this can go in, and is an improvement, if the above is done.

This deadpanned my enthusiasm for this change. I think fixing the above would make this a valuable addition and save people lots of headache, so definitely +1!

**Comment by blink1073** [comment]
I don't like having a separate engine field, since that is too easy to get out of sync.  I think 1) is the way to go.

**Comment by jasongrout** [comment]
> I don't like having a separate engine field, since that is too easy to get out of sync.

Agreed.

**Comment by jasongrout** [comment]
Forgot one more issue (comment inline), and then I think it's good to go in.

File: jupyterlab/labextensions.py [hunk 1]
```diff
@@ -489,18 +491,36 @@ def validate_labextension_folder(name, full_dest, logger=None):
             path = os.path.join(full_dest, fname)
             if not os.path.exists(path):
                 hasFiles = False
-    entry_msg = u"     {} has {} entry point?"
+        # add check here for version compat
+        for (mod, deps) in manifest.get('modules', {}).items():
+            # ignore jupyterlab modules since
+            # they are only semver compatibile with their own version
+            if mod.startswith('jupyterlab'):
```
Comment (jasongrout) [inline jupyterlab/labextensions.py]: Sorry, should be `startswith('jupyterlab@')`

File: jupyterlab/labextensions.py [hunk 2]
```diff
@@ -495,10 +495,10 @@ def validate_labextension_folder(name, full_dest, logger=None):
         for (mod, deps) in manifest.get('modules', {}).items():
             # ignore jupyterlab modules since
             # they are only semver compatibile with their own version
-            if mod.startswith('jupyterlab'):
+            if mod.startswith('jupyterlab/'):
```
Comment (jasongrout) [inline jupyterlab/labextensions.py]: Sorry, I should have said `jupyterlab@` since these are the mangled names.

**Comment by blink1073** [comment]
Kicked the PR so Travis would pick up the commit.

**Comment by jasongrout** [comment]
Thanks!

**Pull request merged**